model/map/src/main/java/org/keycloak/models/map/userSession/MapUserSessionAdapter.java

```diff
@@ -111,22 +111,22 @@ public boolean isOffline() {
         Map<String, AuthenticatedClientSessionModel> result = new HashMap<>();
         List<String> removedClientUUIDS = new LinkedList<>();
 
-        entity.getAuthenticatedClientSessions().entrySet()
-                .stream()
-                .forEach(entry -> {
-                    String clientUUID = entry.getKey();
-                    ClientModel client = realm.getClientById(clientUUID);
-
-                    if (client != null) {
-                        AuthenticatedClientSessionModel clientSession = session.sessions()
-                                .getClientSession(this, client, entry.getValue(), isOffline());
-                        if (clientSession != null) {
-                            result.put(clientUUID, clientSession);
-                        }
-                    } else {
-                        removedClientUUIDS.add(clientUUID);
-                    }
-                });
+        // to avoid concurrentModificationException
```

Comment:
You can reproduce the issue like this:

- revert changes in `MapUserSessionAdapter`
- For example from `UserSessionProviderModelTest#testExpiredClientSessions` in last `inComittedTransaction` call `userSession.getAuthenticatedClientSessions().entrySet().stream().forEach(c -> {});`
- `mvn test -f testsuite/model/pom.xml -Pmap -Dtest=UserSessionProviderModelTest#testExpiredClientSessions`

```
java.util.ConcurrentModificationException
	at java.base/java.util.LinkedHashMap$LinkedHashIterator.nextNode(LinkedHashMap.java:719)
	at java.base/java.util.LinkedHashMap$LinkedEntryIterator.next(LinkedHashMap.java:751)
	at java.base/java.util.LinkedHashMap$LinkedEntryIterator.next(LinkedHashMap.java:749)
	at java.base/java.util.Iterator.forEachRemaining(Iterator.java:133)
	at java.base/java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1801)
	at java.base/java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:658)
	at org.keycloak.models.map.userSession.MapUserSessionAdapter.getAuthenticatedClientSessions(MapUserSessionAdapter.java:116)
```
I'm not sure why it's happening.

Reply:
I found out the reason of the `CME`. The problem is, that `session.sessions().getClientSession()` [udpates](https://github.com/keycloak/keycloak/blob/master/model/map/src/main/java/org/keycloak/models/map/userSession/MapUserSessionProvider.java#L116) `entity.getAuthenticatedClientSessions()` map when the clientSession expires. 

We need to do the `getClientSession` call outside of looping through `authenticatedClientSessions`. I cannot come up with any nice solution, so the copy of `authenticatedClientSessions` is probably the best. But if you can find anything better it would be great.